test/unit/range/iterator_test_template.hpp

```diff
@@ -0,0 +1,518 @@
+    {
+        EXPECT_TRUE(this->begin_iterator  == this->begin_iterator);
+        EXPECT_FALSE(this->begin_iterator != this->begin_iterator);
+    }
+
+    if constexpr (TestFixture::const_iterable)
+    {
+        EXPECT_TRUE(this->begin_const_iterator  == this->begin_const_iterator);
+        EXPECT_FALSE(this->begin_const_iterator != this->begin_const_iterator);
+        EXPECT_FALSE(this->begin_const_iterator == this->end_const_iterator);
+        EXPECT_TRUE(this->begin_const_iterator  != this->end_const_iterator);
+        EXPECT_FALSE(this->end_const_iterator   == this->begin_const_iterator);
+        EXPECT_TRUE(this->end_const_iterator    != this->begin_const_iterator);
+
+        // (non-const lhs)
+        EXPECT_TRUE(this->begin_iterator  == this->begin_const_iterator);
+        EXPECT_FALSE(this->begin_iterator != this->begin_const_iterator);
+        EXPECT_FALSE(this->begin_iterator == this->end_const_iterator);
+        EXPECT_TRUE(this->begin_iterator  != this->end_const_iterator);
+        EXPECT_FALSE(this->end_iterator   == this->begin_const_iterator);
+        EXPECT_TRUE(this->end_iterator    != this->begin_const_iterator);
+
+        // (non-const rhs)
+        EXPECT_TRUE(this->begin_const_iterator  == this->begin_iterator);
+        EXPECT_FALSE(this->begin_const_iterator != this->begin_iterator);
+        EXPECT_FALSE(this->end_const_iterator   == this->begin_iterator);
+        EXPECT_TRUE(this->end_const_iterator    != this->begin_iterator);
+        EXPECT_FALSE(this->begin_const_iterator == this->end_iterator);
+        EXPECT_TRUE(this->begin_const_iterator  != this->end_iterator);
+    }
+}
+
+template <typename test_type, typename it_begin_t, typename it_sentinel_t, typename rng_t>
+inline void move_forward_pre_test(it_begin_t && it_begin, it_sentinel_t && it_end, rng_t && rng)
+{
+    // pre-increment
+    auto rng_it = rng.begin();
+    for (auto it = it_begin; it != it_end; ++it, ++rng_it)
+        expext_eq<test_type>(*it, *rng_it);
+}
+
+template <typename test_type, typename it_begin_t, typename it_sentinel_t, typename rng_t>
+inline void move_forward_post_test(it_begin_t && it_begin, it_sentinel_t && it_end, rng_t && rng)
+{
+    // post-increment
+    auto rng_it = rng.begin();
+    for (auto it = it_begin; it != it_end; it++, ++rng_it)
+        expext_eq<test_type>(*it, *rng_it);
+}
+
+TYPED_TEST_P(iterator_fixture, move_forward_pre)
+{
+    move_forward_pre_test<TypeParam>(this->begin_iterator, this->end_iterator, this->range_to_compare);
+
+    if constexpr (!std::Same<typename TestFixture::iterator_tag, std::input_iterator_tag>) // iterate over it again
+        move_forward_pre_test<TypeParam>(this->begin_iterator, this->end_iterator, this->range_to_compare);
+
+    if constexpr (TestFixture::const_iterable)
+        move_forward_pre_test<TypeParam>(this->begin_const_iterator, this->end_const_iterator, this->range_to_compare);
```

Comment:
Mhmm, I know we are talking about a really rare case (actually not even sure there would be a valid example for it) but it isn't forbidden to have const input iterator. Thus, this would consume twice right? 
Not sure what the most elegant solution would be. Probably by further splitting the test case into const and non-const iterator. Then every fixture is setup again.